Add info about system accounts (#4590)
w3f/polkadot-wiki: docs/learn/learn-account-advanced.md

import RPC from "./../../components/RPC-Connection";

## System Accounts

:::info

You can view system accounts on [Subscan](https://polkadot.subscan.io/account_list?role=module).

:::

> As the word suggests, system accounts are used by the system and do not have users. They are used,

Reviewer: What do you mean, not have users?
```suggestion
As the word suggests, system accounts are used by the system. They are used,
```

> for example, for the treasury, as storage of crowdloan funds, and nomination pools. Those accounts

Reviewer: list is not in the same syntactical form

Author: not sure what you mean here, can you clarify please?

> do not use the normal transfer process, and they will never issue an

Reviewer: what is the "normal" transfer process?

> treasury, as storage of crowdloan funds, and nomination pools. Those accounts they will never issue
> an [extrinsic](./learn-extrinsics.md) since they are not accounts held by users and thus do not have

Reviewer 2: ```suggestion
treasury, as storage of crowdloan funds, and nomination pools. These accounts will never issue
an [extrinsic](./learn-extrinsics.md) since they are not accounts held by users and thus do not have
```

> [extrinsic](./learn-extrinsics.md) since they are not "real" accounts held by users.

Reviewer: they are very real accounts

> The `Account` column in the list of system accounts on

Reviewer 2: ```suggestion
The `Account` column in the table of system accounts on
```

[Subscan](https://polkadot.subscan.io/account_list?role=module) shows the ID for each system

> account. Some of those ID are Substrate pallet IDs. For example, the treasury pallet id is

Reviewer 2: ```suggestion
account. Some of those IDs are Substrate pallet IDs. For example, the treasury pallet ID is
```

[`py/trsry`](https://github.com/paritytech/polkadot/blob/6282def1bb053858522cf551b86b2d07aad04f29/runtime/polkadot/src/lib.rs#L818).

> It is possible to derive account addresses from these pallet IDs in the same way parachain IDs can

Reviewer 2: ```suggestion
It is possible to derive account addresses from these pallet IDs. In the same way, parachain IDs can
```

can also be represented as addresses. For example, using the
[substrate js utilities](https://www.shawntabrizi.com/substrate-js-utilities/) by Shawn Tabrizi,
converting `py/trsry -> address` will return the substrate address for the treasury:

`5EYCAe5ijiYfyeZ2JJCGq56LmPyNRAKzpG4QkoQkkQNB5e6Z`

This acts as a container for funds but has no actual power over them due to a lack of a

> corresponding key pair. Given the multi-chain property of the ecosystem, such address can be

Reviewer: Does any address have actual power?

> mirrored to all parachains so that each parachain has its treasury. At the time of writing, on

Reviewer: There's no inherent reason that that address would have treasury _logic_ associated with it on a particular chain.

it on those networks.

Different types of system accounts also include:

- `py/cfund` for crowdloans (Substrate pallet ID)

> - `X` for parathreads where X is the parathread number

Reviewer: these are wrong

Author: in Tabrizi's Tool :

if I type in Module ID `py/cfund` I get `5EYCAe5ijiYdg22N9CWUvWYfa7De8WmxXdP78KZxhy7N91uv` --> `modlpy/cfund:0`

if I type the number `2000` in the para ID field, I get `5Ec4AhPUwPeyTFyuhGuBbD224mY85LKLMSqSSo33JYWCazU4` --> `para:2000`

any advice on how can I address your feedback? Is the wording wrong?

Reviewer 2: @joepetrowski I changed the content based on https://substrate.stackexchange.com/questions/536/how-do-treasury-accounts-compare-to-end-user-accounts-in-frame and kept it simple. Let us know if the content in the PR now is factually correct.

:::

As the word suggests, system accounts are used by the system and do not have users. They are used,
for example, for the treasury, as storage of crowdloan funds, and nomination pools. Those accounts
do not use the normal transfer process, and they will never issue an
[extrinsic](./learn-extrinsics.md) since they are not "real" accounts held by users.

The `Account` column in the list of system accounts on

> [Subscan](https://polkadot.subscan.io/account_list?role=module) shows the substrate pallet ID for

Reviewer 3: ```suggestion
[Subscan](https://polkadot.subscan.io/account_list?role=module) shows the Substrate pallet ID for
```

I'm not entirely sure if *each* one is a pallet id, we would need more info on this.

Reviewer: indeed, they are not

each system account. For example, the treasury pallet id is
[`py/trsry`](https://github.com/paritytech/polkadot/blob/6282def1bb053858522cf551b86b2d07aad04f29/runtime/polkadot/src/lib.rs#L818).

It is possible to derive account IDs/addresses from these pallet IDs in the same way parachain IDs
can also be represented as addresses. For example, using the

> [substrate js utilities](https://www.shawntabrizi.com/substrate-js-utilities/) by Shawn Tabrizi,

Reviewer 3: ```suggestion
[Substrate JS Utilities tool](https://www.shawntabrizi.com/substrate-js-utilities/) by Shawn Tabrizi,
```

> converting `py/trsry -> address` will return the substrate address for the treasury:

Reviewer 3: ```suggestion
converting `py/trsry -> address` will return the Substrate address for the treasury:
```

counterintuitive that some addresses always have the same prefix and others like Kusama can vary
wildly, but it's just a quirk of Base58-check encoding.

## System Accounts

> 

Reviewer: ```suggestion
```

As the word suggests, system accounts are used by the system. They are used, for example, for the
treasury, crowdloans, and nomination pools. From the point of view of the runtime, these accounts 
are like any other account on-chain. These special system accounts are just public keys, with the 
private key being unknown (and unattainable). So, that means that only the pallet itself can interact 
with this account. These accounts can never issue a signed [extrinsic](./learn-extrinsics.md) since they 

> do not have a private key.

Reviewer: I don't understand the obsession with public key cryptography that kind of runs throughout all this. It really distracts from what an account actually is. Accounts are just storage items in the system that are identified by 32-byte keys, and have values that conform to a defined datagram. There are lots of ways to come up with 32-byte keys, like coming up with a word, let's say "py/trsry" and zero-padding it to be 32 bytes. "polkadotpolkadotpolkadotpolkadot" is also 32 bytes (that is 5Ec8KWYgo6B4xTjtu6jo2NZz12ouH1VNqJPcC9rvqiscERnu). You can send DOT to that address. And you can spend it, via governance.

Of course, I can generate a secret, and derive a 32 byte value that is safe to share with others without exposing any information about the secret. And I can generate messages (signatures) using that secret to show that I know the secret from which that value was derived without revealing any information about the secret. Don't get me wrong, this is cool. But irrelevant.

There are lots of ways to create rules about when the value associated with a key can be mutated. Of course one of them is signing messages. Another one is via the very logic that created the key and made it "special". Another is via governance, which can mutate the value behind any key, whether it was generated as the counterpart of a secret, as part of some deterministic logic, or was just nonsense.

Author: Thank you Joe, there is a nice definition of Account here that we can add into the wiki in my opinion. Maybe at the beginning of the page, @DrW3RK what do you think?

Reviewer 2: I am fine with removing the whole account cryptography part. I only tried to make the existing content in the PR more intuitive. Will do that now

:::info

You can view system accounts on [Subscan](https://polkadot.subscan.io/account_list?role=module).

:::

For instance, let us take a look at how treasury accounts are created and used under the hood. 
To generate the treasury account, the raw bytes of the strings "modl" and "py/trsry" are combined
to create the `AccountID`

```
as_bytes("modl") = 0x6d6f646c
as_bytes("py/trsry") = 0x70792f7472737279
```

> Then the rest of the account address is filled with 0s till it reaches the expected account encoded length.

Reviewer: ```suggestion
Then the rest of the account address is filled with zeros until it reaches the expected account encoded length.
```

```
0x6d6f646c70792f74727372790000000000000000000000000000000000000000
```

Which maps to Substrate Address:

```
5EYCAe5ijiYfyeZ2JJCGq56LmPyNRAKzpG4QkoQkkQNB5e6Z
```

For more information, check the post on Substrate StackExchange on [Treasury accounts](https://substrate.stackexchange.com/questions/536/how-do-treasury-accounts-compare-to-end-user-accounts-in-frame).

Similarly, the crowdloan system accounts are derived from the pallet ID as well as the fund account index.
To generate the crowdloan account, the raw bytes of the strings "modl" and "py/cfund" along with the fund index
are combined to create the `AccountID`. For instance, "modlpy/cfund:0" represents the first ever crowdloan account

> created. For all the subsequent crowdloans, the index keeps incrementing. Similar logic applies to Nomination Pools
> and Parachain accounts as well.

Reviewer: ```suggestion
created. For all the subsequent crowdloans, the index keeps incrementing. Similar logic applies to nomination pool
and parachain accounts as well.
```
